File: Lua/bitonic_sort.lua
function bitonic_compare(arr, low, cnt, dir)
    if cnt > 1 then
        local k = cnt // 2
        for i = low, low + k - 1 do
            if (dir and arr[i] > arr[i + k]) or (not dir and arr[i] < arr[i + k]) then
                arr[i], arr[i + k] = arr[i + k], arr[i]
            end
        end
        bitonic_compare(arr, low, k, dir)
        bitonic_compare(arr, low + k, k, dir)
    end
end

function bitonic_merge(arr, low, cnt, dir)
    if cnt > 1 then
        local k = cnt // 2
        bitonic_merge(arr, low, k, true)
        bitonic_merge(arr, low + k, k, false)
        bitonic_compare(arr, low, cnt, dir)
    end
end

function bitonic_sort(arr, low, cnt, dir)
    if cnt > 1 then
        local k = cnt // 2
        bitonic_sort(arr, low, k, true)
        bitonic_sort(arr, low + k, k, false)
        bitonic_merge(arr, low, cnt, dir)
    end
end

local arr = {3, 7, 4, 8, 6, 2, 1, 5}
bitonic_sort(arr, 1, #arr, true)
print(table.concat(arr, " "))

File: C/bitonic_sort.c
#include <stdio.h>

#define ASCENDING 1
#define DESCENDING 0

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void bitonicMerge(int arr[], int low, int cnt, int dir) {
    if (cnt > 1) {
        int k = cnt / 2;
        for (int i = low; i < low + k; i++) {
            if ((dir == ASCENDING && arr[i] > arr[i + k]) ||
                (dir == DESCENDING && arr[i] < arr[i + k])) {
                swap(&arr[i], &arr[i + k]);
            }
        }
        bitonicMerge(arr, low, k, dir);
        bitonicMerge(arr, low + k, k, dir);
    }
}

void bitonicSort(int arr[], int low, int cnt, int dir) {
    if (cnt > 1) {
        int k = cnt / 2;
        bitonicSort(arr, low, k, ASCENDING);
        bitonicSort(arr, low + k, k, DESCENDING);
        bitonicMerge(arr, low, cnt, dir);
    }
}

void sort(int arr[], int n) {
    bitonicSort(arr, 0, n, ASCENDING);
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[] = {3, 7, 2, 5, 8, 4, 6, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    sort(arr, n);
    printf("Sorted array: \n");
    printArray(arr, n);

    return 0;
}